File: ex/hexagonal/src/main/kotlin/com/ktb/shop/infra/email/LogNotifier.kt
package com.ktb.shop.infra.email

import com.ktb.shop.domain.Customer
import com.ktb.shop.domain.port.Notifier
import jakarta.inject.Singleton
import org.slf4j.LoggerFactory

// ADAPTER for the Notifier port. Swap for a real SMTP/SendGrid adapter later —
// the core (RegisterCustomer) does not change.
@Singleton
class LogNotifier : Notifier {
    private val log = LoggerFactory.getLogger(javaClass)

    override fun welcome(customer: Customer) {
        log.info("welcome email -> {} ({})", customer.email.value, customer.name)
    }
}

File: ex/hexagonal-go/internal/infra/email/log_notifier.go
package email

import (
	"context"
	"log/slog"

	"github.com/ktb/hexagonal-shop/internal/domain"
)

// ADAPTER for the Notifier port. Swap for a real SMTP/SendGrid adapter later —
// the core (RegisterCustomer) does not change.
type LogNotifier struct{}

func NewLogNotifier() *LogNotifier { return &LogNotifier{} }

func (n *LogNotifier) Welcome(_ context.Context, customer domain.Customer) {
	slog.Info("welcome email", "to", customer.Email.String(), "name", customer.Name)
}
